mycli/output_formatter.py

```diff
@@ -0,0 +1,232 @@
+    """
+    quote = len(data[0]) * [False]
+    for row in data:
+        for i, v in enumerate(row):
+            v = encodingutils.text_type(v)
+            if v.startswith(' ') or v.endswith(' '):
+                quote[i] = True
+
+    results = []
+    for row in data:
+        result = []
+        for i, v in enumerate(row):
+            quotation = quotestyle if quote[i] else ''
+            result.append('{quotestyle}{value}{quotestyle}'.format(
+                quotestyle=quotation, value=v))
+        results.append(result)
+    return results, headers
+
+
+def tabulate_wrapper(data, headers, table_format=None, missing_value='', **_):
+    """Wrap tabulate inside a standard function for OutputFormatter."""
+    return tabulate(data, headers, tablefmt=table_format,
+                    missingval=missing_value, disable_numparse=True)
+
+
+def csv_wrapper(data, headers, delimiter=',', **_):
+    """Wrap CSV formatting inside a standard function for OutputFormatter."""
+    with contextlib.closing(StringIO()) as content:
+        writer = csv.writer(content, delimiter=str(delimiter))
+
+        writer.writerow(headers)
+        for row in data:
+            writer.writerow(row)
+
+        return content.getvalue()
+
+
+def terminal_tables_wrapper(data, headers, table_format=None, **_):
+    """Wrap terminaltables inside a standard function for OutputFormatter."""
+    if table_format == 'ascii':
+        table = terminaltables.AsciiTable
+    elif table_format == 'single':
+        table = terminaltables.SingleTable
+    elif table_format == 'double':
+        table = terminaltables.DoubleTable
+    elif table_format == 'github':
+        table = terminaltables.GithubFlavoredMarkdownTable
+    else:
+        raise ValueError('unrecognized table format: {}'.format(table_format))
+
+    t = table([headers] + data)
+    return t.table
+
+
+class OutputFormatter(object):
+    """A class with a standard interface for various formatting libraries."""
+
+    def __init__(self, format_name=None):
+        """Register the supported output formats."""
+        self._output_formats = {
+            'csv': (csv_wrapper, {
+                'preprocessor': (override_missing_value, bytes_to_string),
+                'missing_value': '<null>'
+            }),
+            'tsv': (csv_wrapper, {
+                'preprocessor': (override_missing_value, bytes_to_string),
+                'missing_value': '<null>',
+                'delimiter': '\t'
+            }),
+            'expanded': (expanded_table, {
+                'preprocessor': (override_missing_value, convert_to_string),
+                'missing_value': '<null>'
+            })
+        }
+        self._format_name = None
+
+        tabulate_formats = ('plain', 'simple', 'grid', 'fancy_grid', 'pipe',
+                            'orgtbl', 'jira', 'psql', 'rst', 'mediawiki',
+                            'moinmoin', 'html', 'latex', 'latex_booktabs',
+                            'textile')
```

Comment:
Adding module specific code into OutputFormatter makes this class extremely specialized. We should try to avoid tight coupling as much as possible. 

Since OutputFormatter is supposed to be an interface, we should make this completely agnostic to the implentation of the formatter itself.